sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,195 @@
+
+The plugin's main class is inherited from basic class ``Layer``:
+
+.. code-block:: c
+
+  class GradientLayer : public Layer
+
+It must have implementations at least for the following methods: ``updateBounds()``, ``updateCosts()`` and ``reset()``. However, in this example for better comprehensibility of Costmap2D plugin structure the following methods were overriden: ``onInitialize()``, ``updateBounds()``, ``updateCosts()``, ``onFootprintChanged()``, ``reset()``.
+
+First 4 methods have concrete functionalities:
+
+1. ``GradientLayer::onInitialize()`` method is called at the end of plugin initialization. There is usually declaration(s) of ROS parameter(s) with its default value(s):
+
+.. code-block:: c
+
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  node_->get_parameter(name_ + "." + "enabled", enabled_);
+
+In current implementation there is also setting ``need_recalculation_`` bounds recalculation indicator:
+
+.. code-block:: c
+
+   need_recalculation_ = false;
+
+2. ``GradientLayer::updateCosts()`` method is called by the ``LayeredCostmap`` each time when costmap re-calculation is required. It updates the costmap layer only within its bounds. There are 4 input parameters of method: calculation window bounds and 1 output parameter: reference to ``GradientLayer`` costmap. Inside this method it is placed the gradient making algorithm for underlying costmap:
+
+.. code-block:: c
+
+  int gradient_index;
+  for (int j = min_j; j < max_j; j++) {
+    // Reset gradient_index each time when reaching the end of re-calculated window
+    // by OY axis.
+    gradient_index = 0;
+    for (int i = min_i; i < max_i; i++) {
+      int index = master_grid.getIndex(i, j);
+      // assign the cost associated with the distance from an obstacle to the cell
+      unsigned char cost = (LETHAL_OBSTACLE - gradient_index*GRADIENT_FACTOR)%255;
+      if (gradient_index <= GRADIENT_SIZE) {
+        gradient_index++;
+      } else {
+        gradient_index = 0;
+      }
+      master_array[index] = cost;
+    }
+  }
+
+where the ``GRADIENT_SIZE`` is the size of each gradient period in map cells, ``GRADIENT_FACTOR`` - decrement of costmap's value per each step:
+
+.. image:: images/Writing_new_Costmap2D_plugin/gradient_explanation.png
+
+These parameters are defined in plugin's header file.
+
+3. ``GradientLayer::updateBounds()`` is called by the ``LayeredCostmap`` to ask the plugin: which area of costmap layer it needs to update. There are 3 input parameters of method: robot position and orientation and 4 output parameters: pointers to window bounds. In current implementation ``updateBounds()`` re-calculates window bounds if ``need_recalculation_`` is ``true`` and updates them independently on ``need_recalculation_`` value.
```

Comment:
This should be #2 since it is called before the updateCosts function, we should go in order